### PR TITLE
Move ERC-7739 to Validators

### PR DESCRIPTION
**Pull request by filmakarov**
Moved ERC-7739 from Nexus to validators (K1, Mock)

- Added ERC7739Validator base contract
- 712 domain is taken from smart account in a validator when calculating hashes
- Added `supportsNestedTypedDataSign() ` that iterates over validators

**TODO:** 
- [x] test supportsNestedTypedDataSign
- [x] fix tests for k1 and enable mode that fail because validators now expect 7739 sigs
- [x] Optimize hashing name and version in ERC7739Validator
- [x] Add test with module enable mode for account deployment userOp

### CODE REVIEW

File: contracts/Nexus.sol [hunk 1]
```diff
@@ -195,16 +194,15 @@ contract Nexus is INexus, BaseAccount, ExecutionHelper, ModuleManager, UUPSUpgra
 
     /// @notice Validates a signature according to ERC-1271 standards.
     /// @param hash The hash of the data being validated.
-    /// @param data Signature data that needs to be validated.
+    /// @param signature Signature data that needs to be validated.
     /// @return The status code of the signature validation (`0x1626ba7e` if valid).
     /// bytes4(keccak256("isValidSignature(bytes32,bytes)") = 0x1626ba7e
     /// @dev Delegates the validation to a validator module specified within the signature data.
-    function isValidSignature(bytes32 hash, bytes calldata data) external view virtual override returns (bytes4) {
+    function isValidSignature(bytes32 hash, bytes calldata signature) external view virtual override returns (bytes4) {
```
Comment (livingrockrises) [inline contracts/Nexus.sol:201]: why rename this signature over generic data?
typically signature is extracted way later

File: contracts/Nexus.sol [hunk 2]
```diff
@@ -291,196 +289,43 @@ contract Nexus is INexus, BaseAccount, ExecutionHelper, ModuleManager, UUPSUpgra
...
+    /// for discussions.
+    ///
     /// By default, it returns `bytes32(bytes4(keccak256("supportsNestedTypedDataSign()")))`,
     /// denoting support for the default behavior, as implemented in
     /// `_erc1271IsValidSignatureViaNestedEIP712`, which is called in `isValidSignature`.
     /// Future extensions should return a different non-zero `result` to denote different behavior.
     /// This method intentionally returns bytes32 to allow freedom for future extensions.
-    function supportsNestedTypedDataSign() public pure virtual returns (bytes32 result) {
-        result = bytes4(0xd620c85a);
+    function supportsNestedTypedDataSignWithValidator(address validator) public view virtual returns (bytes32) {
```
Comment (livingrockrises) [inline contracts/Nexus.sol:304]: so we pass the validator and ask the account if this validator supports 7739? 
why can't read method just sit in the validator? is there any other usage of this

File: contracts/Nexus.sol [hunk 3]
```diff
@@ -291,196 +289,43 @@ contract Nexus is INexus, BaseAccount, ExecutionHelper, ModuleManager, UUPSUpgra
...
+        return(IERC7739(validator).supportsNestedTypedDataSign());
+    }
+
+    /// @dev For automatic detection that the smart account supports the nested EIP-712 workflow
+    /// Offchain usage only
+    /// Iterates over all the validators
+    function supportsNestedTypedDataSign() public view virtual returns (bytes32) {
```
Comment (livingrockrises) [inline contracts/Nexus.sol:311]: this make sense - eventually it's just the job of the validator.

Comment (livingrockrises) [inline contracts/Nexus.sol:311]: return should be bytes4

File: contracts/Nexus.sol [hunk 4]
```diff
@@ -291,196 +289,43 @@ contract Nexus is INexus, BaseAccount, ExecutionHelper, ModuleManager, UUPSUpgra
...
-    }
-
     /// @dev EIP712 domain name and version.
     function _domainNameAndVersion() internal pure override returns (string memory name, string memory version) {
         name = "Nexus";
         version = "1.0.0-beta";
     }
-
-    /// @dev For use in `_erc1271HashForIsValidSignatureViaNestedEIP712`,
-    function _typedDataSignFields() private view returns (bytes32 m) {
```
Comment (livingrockrises) [inline contracts/Nexus.sol:462]: thank god

File: contracts/base/ERC7739Validator.sol [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    bytes32 internal constant _DOMAIN_TYPEHASH = 0x8b73c3c69bb8fe3d512ecc4cf759cc79239f7b179b0ffacaa9a75d522b39400f;
+    
+    /// @dev For automatic detection that the smart account supports the nested EIP-712 workflow.
+    /// By default, it returns `bytes32(bytes4(keccak256("supportsNestedTypedDataSign()")))`,
+    /// denoting support for the default behavior, as implemented in
+    /// `_erc1271IsValidSignatureViaNestedEIP712`, which is called in `isValidSignature`.
+    /// Future extensions should return a different non-zero `result` to denote different behavior.
+    /// This method intentionally returns bytes32 to allow freedom for future extensions.
+    function supportsNestedTypedDataSign() public view virtual returns (bytes32 result) {
+        result = bytes4(0xd620c85a);
```
Comment (livingrockrises) [inline contracts/base/ERC7739Validator.sol:23]: more bytes4 hexes :) go 7739

File: contracts/base/ERC7739Validator.sol [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    /// - https://github.com/junomonster/nested-eip-712
+    /// - https://github.com/frangio/eip712-wrapper-for-eip1271
+    ///
+    /// Their nomenclature may differ from ours, although the high-level idea is similar.
+    ///
+    /// Of course, if you have control over the codebase of the wallet client(s) too,
+    /// you can choose a more minimalistic signature scheme like
+    /// `keccak256(abi.encode(address(this), hash))` instead of all these acrobatics.
+    /// All these are just for widespread out-of-the-box compatibility with other wallet clients.
+    function _erc1271HashForIsValidSignatureViaNestedEIP712(
```
Comment (livingrockrises) [inline contracts/base/ERC7739Validator.sol:94]: ok can you point me to the tests written on this in smart session validator please

File: test/foundry/unit/concrete/erc1271/TestERC1271Account_IsValidSignature.t.sol [hunk 7]
```diff
@@ -144,14 +145,16 @@ contract TestERC1271Account_IsValidSignature is NexusTest_Base {
     /// @notice Generates an ERC-1271 hash for personal sign.
     /// @param childHash The child hash.
     /// @return The ERC-1271 hash for personal sign.
-    function toERC1271HashPersonalSign(bytes32 childHash) internal view returns (bytes32) {
+    function toERC1271HashPersonalSign(bytes32 childHash, address account) internal view returns (bytes32) {
```
Comment (livingrockrises) [inline test/foundry/unit/concrete/erc1271/TestERC1271Account_IsValidSignature.t.sol:148]: these tests must be broken right?

**Review by livingrockrises — COMMENTED** [review]
review i
might need some help in debugging tests #PTSD

**Comment by livingrockrises** [comment]
needs to resolve conflicts too.

File: contracts/Nexus.sol (hunk 3 above)
Comment (filmakarov) [reply contracts/Nexus.sol:340]: nope. it's bytes32 
see here https://ethereum-magicians.org/t/erc-7739-readable-typed-signatures-for-smart-accounts/20513#supportsnestedtypeddatasign-function-for-detection-16

File: contracts/Nexus.sol (hunk 1 above)
Comment (filmakarov) [reply contracts/Nexus.sol:230]: just want it to be closer to 1271 spec

**Comment by filmakarov** [comment]
> review i might need some help in debugging tests #PTSD

all tests pass. no debug needed. 
all conflicts so far resolved as well I believe

**Review by livingrockrises — APPROVED** [review]

File: test/foundry/unit/concrete/erc1271/TestERC1271Account_MockProtocol.t.sol [hunk 8]
```diff
@@ -152,7 +156,7 @@ contract TestERC1271Account_MockProtocol is NexusTest_Base {
     /// @param contents The contents hash.
     /// @param account The address of the account.
     /// @return The ERC-1271 hash.
-    function toERC1271Hash(bytes32 contents, address payable account) internal view returns (bytes32) {
+    function toERC1271Hash(bytes32 contents, address account) internal view returns (bytes32) {
```
Comment (Aboudjem) [inline test/foundry/unit/concrete/erc1271/TestERC1271Account_MockProtocol.t.sol:159]: internal should have `_` as prefix

File: contracts/mocks/MockValidator.sol [hunk 9]
```diff
@@ -2,36 +2,50 @@
...
+contract MockValidator is ERC7739Validator, IERC1271Unsafe {
     mapping(address => address) public smartAccountOwners;
 
     function validateUserOp(PackedUserOperation calldata userOp, bytes32 userOpHash) external view returns (uint256 validation) {
-        return
-            ECDSA.recover(MessageHashUtils.toEthSignedMessageHash(userOpHash), userOp.signature) == smartAccountOwners[msg.sender]
-                ? VALIDATION_SUCCESS
-                : VALIDATION_FAILED;
+        address owner = smartAccountOwners[msg.sender];
+        return _validateSignatureForOwner(owner, userOpHash, userOp.signature) ? VALIDATION_SUCCESS : VALIDATION_FAILED;
```
Comment (Aboudjem) [inline contracts/mocks/MockValidator.sol:19]: owner can be inlined into the return to save some gas

File: contracts/interfaces/modules/IERC1271Unsafe.sol [hunk 10]
```diff
@@ -0,0 +1,15 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.27;
+
+interface IERC1271Unsafe {
```
Comment (Aboudjem) [inline contracts/interfaces/modules/IERC1271Unsafe.sol:4]: Can we rename this? "Unsafe" gives a bad signal. Maybe something like IERC1271Legacy or similar would be better

File: contracts/Nexus.sol (hunk 2 above)
Comment (livingrockrises) [reply contracts/Nexus.sol:334]: it's fine. we wont run into codesize problems now. and its implementation contract deployed once so ok

File: contracts/interfaces/modules/IERC1271Unsafe.sol (hunk 10 above)
Comment (filmakarov) [reply contracts/interfaces/modules/IERC1271Unsafe.sol:4]: I like legacy, yes

File: contracts/mocks/MockValidator.sol (hunk 9 above)
Comment (filmakarov) [reply contracts/mocks/MockValidator.sol:19]: inlined for K1
for Mock left as is for verbosity as it isn't going to prod

File: contracts/base/ModuleManager.sol [hunk 11]
```diff
@@ -379,32 +382,43 @@ abstract contract ModuleManager is Storage, EIP712, IModuleManagerEventsAndError
...
+        // have to try different ways to validate the signature as we do not know for sure what 1271 flavours validator supports
+        if (IERC7739(enableModeSigValidator).supportsNestedTypedDataSign() == SUPPORTS_NESTED_TYPED_DATA_SIGN) {
+            // if the validator supports 7739, we use just the struct hash, as the full hash will be rebuilt inside 7739 flow
+            if (IValidator(enableModeSigValidator).isValidSignatureWithSender(address(this), eip712Digest, sig[20:]) == ERC1271_MAGICVALUE) {
+                return true;
+            }
+        } else {
+            // if the validator doesn't support 7739 for standard isValidSignatureWithSender, we provide the eip 712 digest
+            if (IValidator(enableModeSigValidator).isValidSignatureWithSender(address(this), eip712Digest, sig[20:]) == ERC1271_MAGICVALUE) {
+                return true;
```
Comment (NoahMarconi) [inline contracts/base/ModuleManager.sol:402]: are these two code blocks meant to be identical?

Comment (filmakarov) [reply contracts/base/ModuleManager.sol:402]: nope. great catch. I just noticed this today morning and fixed in #169

**Pull request merged**